test/pilosa.go

```diff
@@ -8,41 +8,114 @@ import (
 	"testing"
 
 	"github.com/pilosa/pilosa/server"
+	"github.com/pkg/errors"
 )
 
 func MustNewRunningServer(t *testing.T) *server.Command {
+	s, err := newServer()
+	if err != nil {
+		t.Fatalf("getting new server: %v", err)
+	}
+
+	err = s.Run()
+	if err != nil {
+		t.Fatalf("running new pilosa server: %v", err)
+	}
+	return s
+}
+
+func newServer() (*server.Command, error) {
 	s := server.NewCommand(&bytes.Buffer{}, ioutil.Discard, ioutil.Discard)
-	s.Config.Bind = ":0"
-	port := strconv.Itoa(MustOpenPort(t))
-	s.Config.GossipPort = port
-	s.Config.GossipSeed = "localhost:" + port
+
+	port, err := openPort()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting port")
+	}
+	s.Config.Bind = "localhost:" + strconv.Itoa(port)
+
+	gport, err := openPort()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting gossip port")
+	}
+	s.Config.GossipPort = strconv.Itoa(gport)
+
+	s.Config.GossipSeed = "localhost:" + s.Config.GossipPort
+	s.Config.Cluster.Type = "gossip"
+	s.Config.Metric.Diagnostics = false
 	td, err := ioutil.TempDir("", "")
 	if err != nil {
-		t.Fatalf("error creating temp data directory: %v", err)
+		return nil, errors.Wrap(err, "temp dir")
 	}
 	s.Config.DataDir = td
-	err = s.Run()
+	return s, nil
+}
+
+func openPort() (int, error) {
```

Comment:
I think the name of this function is a bit misleading. As far as I see it finds the first available port. Also can another process start listening the returned port before the server does in the `newServer` function?

Reply:
Wouldn't just setting the listen address to `localhost:0` and retrieve the port later work better?

Reply:
e.g., `s.Config.Bind = "localhost:0`

Reply:
We need to give every host in the cluster a gossip seed before we start them, and we need to configure each host with the list of hosts in the cluster before we start them. There are certainly changes we could make, but I think the cluster-resize work will address a lot of these issues, so I'm hesitant to do anything too invasive.